src/PolynomialTransform.cc

```diff
@@ -0,0 +1,280 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#include "lsst/meas/astrom/PolynomialTransform.h"
+#include "lsst/meas/astrom/SipTransform.h"
+#include "lsst/meas/astrom/detail/polynomialUtils.h"
+
+namespace lsst { namespace meas { namespace astrom {
+
+PolynomialTransform PolynomialTransform::convert(ScaledPolynomialTransform const & scaled) {
+    return compose(
+        scaled.getOutputScalingInverse(),
+        compose(scaled.getPoly(), scaled.getInputScaling())
+    );
+}
+
+PolynomialTransform PolynomialTransform::convert(SipForwardTransform const & other) {
+    PolynomialTransform poly = other.getPoly();
+    poly._xCoeffs(1, 0) += 1;
+    poly._yCoeffs(0, 1) += 1;
+    return compose(
+        other.getCDMatrix(),
+        compose(
+            poly,
+            afw::geom::AffineTransform(afw::geom::Point2D() - other.getPixelOrigin())
+        )
+    );
+}
+
+PolynomialTransform PolynomialTransform::convert(SipReverseTransform const & other) {
+    PolynomialTransform poly = other.getPoly();
+    poly._xCoeffs(1, 0) += 1;
+    poly._yCoeffs(0, 1) += 1;
+    return compose(
+        afw::geom::AffineTransform(afw::geom::Extent2D(other.getPixelOrigin())),
+        compose(
+            poly,
+            other._cdInverse
+        )
+    );
+}
+
+PolynomialTransform::PolynomialTransform(int order) :
+    _xCoeffs(ndarray::allocate(order + 1, order + 1)),
+    _yCoeffs(ndarray::allocate(order + 1, order + 1)),
+    _u(order + 1),
+    _v(order + 1)
+{
+    if (order < 0) {
+        throw LSST_EXCEPT(
+            pex::exceptions::LengthError,
+            "PolynomialTransform order must be >= 0"
+        );
+    }
+    _xCoeffs.setZero();
+    _yCoeffs.setZero();
+}
+
+PolynomialTransform::PolynomialTransform(
+    ndarray::Array<double const,2,2> const & xCoeffs,
+    ndarray::Array<double const,2,2> const & yCoeffs
+) : _xCoeffs(ndarray::copy(xCoeffs)),
+    _yCoeffs(ndarray::copy(yCoeffs)),
+    _u(_xCoeffs.rows()),
+    _v(_xCoeffs.rows())
+{
+    if (xCoeffs.getShape() != yCoeffs.getShape()) {
+        throw LSST_EXCEPT(
+            pex::exceptions::LengthError,
+            "X and Y coefficient matrices must have the same shape."
+        );
+    }
+    if (_xCoeffs.cols() != _xCoeffs.rows()) {
+        throw LSST_EXCEPT(
+            pex::exceptions::LengthError,
+            "Coefficient matrices must be triangular, not trapezoidal."
```

Comment:
For matrices, trapezoidal is to triangular as rectangular is to square.  They're triangular because we don't want nonzero terms with p+q > order, though it's not worth out time to test that.

Offending values printed.